qubes-gui-client/src/buffer.rs

```diff
@@ -267,30 +264,24 @@ impl<T: VchanMock> RawMessageStream<T> {
                             break Err(Error::new(ErrorKind::InvalidData, format!("{}", e)));
                         }
                         Ok(Some(header)) => {
-                            // length sanitized by validate_length()
-                            self.buffer.resize(header.len(), 0);
-                            break process_so_far(self, header, ready - size_of::<Header>(), 0);
+                            break process_so_far(self, header, ready - size_of::<Header>());
```

Comment:
Couldn't you just set `self.state` to `ReadingBody` here, and let the loop this is in run that part of the code for you? That would remove the need for the `process_so_far` closure.